[vcpkg scripts] Adding support for visionOS

colincornaby opened the pull request:
This change adds visionOS support. It does not modify any ports and does not guarantee port compatibility. (And some ports would certainly need some fix ups in the future.)

visionOS is a variant of iOS so most of the existing support can just be copied over. The visionOS SDK is still different so it's not sufficient just to use iOS builds.

This PR does not attempt to get the visionOS simulator working. The iOS scripts do try to enable simulator support - but I think that support is not currently functioning. They do a check for x86 or x86_64 processor architectures to decide if a simulator build needs to occur. But the visionOS simulator only runs on a ARM64 host so a processor architecture check will not be sufficient.

This commit requires https://github.com/microsoft/vcpkg-tool/pull/1681.

colincornaby committed:
Adding support for visionOS

JoergAtGithub commented on scripts/cmake/vcpkg_common_definitions.cmake:
```diff
@@ -69,7 +71,7 @@ else()
 endif()
 
-if(VCPKG_TARGET_IS_OSX OR VCPKG_TARGET_IS_IOS)
+if(VCPKG_TARGET_IS_OSX OR VCPKG_TARGET_IS_IOS OR VCPKG_TARGET_IS_VCPKG)
```
This looks like a typo

Cheney-W replied:
@colincornaby Would it be appropriate to use VCPKG_TARGET_IS_VISIONOS here?

BillyONeal commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -352,6 +352,8 @@ elseif(CMAKE_SYSTEM_NAME STREQUAL "Darwin" OR (NOT CMAKE_SYSTEM_NAME AND CMAKE_H
 elseif(CMAKE_SYSTEM_NAME STREQUAL "iOS")
     set(Z_VCPKG_TARGET_TRIPLET_PLAT ios)
+elseif(CMAKE_SYSTEM_NAME STREQUAL "visionOS")
+    set(Z_VCPKG_TARGET_TRIPLET_PLAT visionos)
```
There need to be triplets that actually have this name for this detection to make sense.

Note for other reviewers: It's OK that there is no detection for this from MSBuild because MSBuild won't ever target visionOS.

BillyONeal replied:
Oh I see that one is added here. I keep missing stuff reviewing this, sorry 😅

colincornaby commented on scripts/toolchains/visionos.cmake:
```diff
@@ -0,0 +1,69 @@
+if(NOT _VCPKG_VISIONOS_TOOLCHAIN)
```
I'll investigate reusing the iOS one.

colincornaby commented on scripts/toolchains/ios.cmake:
```diff
@@ -1,5 +1,9 @@
 if(NOT _VCPKG_IOS_TOOLCHAIN)
-    set(_VCPKG_IOS_TOOLCHAIN 1)
+    if ("$VCPKG_CMAKE_SYSTEM_NAME" STREQUAL "visionOS")
```
I'm not sure if this is the best way to check the platform within a toolchain. VCPKG_TARGET_IS_VISIONOS seems like another option if it's already defined before the toolchain is loaded.

BillyONeal replied:
I don't think any kind of detection is intended here; `_VCPKG_IOS_TOOLCHAIN` is merely a double-inclusion guard. I think this whole block should be reverted.

colincornaby commented on scripts/toolchains/ios.cmake:
```diff
@@ -57,7 +64,7 @@ if(NOT _VCPKG_IOS_TOOLCHAIN)
     # so the following will be no-ops.
     set(CMAKE_OSX_ARCHITECTURES "${_vcpkg_ios_target_architecture}" CACHE STRING "Build architectures for iOS")
-    if(_vcpkg_ios_sysroot)
+    if(_vcpkg_ios_sysroot AND NOT _VCPKG_IOS_TOOLCHAIN)
```
I'm opting not to set a sysroot if the target is visionOS. It shouldn't be necessary unless the target is a simulator. But the previous author explicitly defined the value so if there is a reason I should I can do so. visionOS has a sysroot of `xros` so I'd need to break the case out.

BillyONeal replied:
This has been here since the initial addition of iOS:

https://github.com/microsoft/vcpkg/commit/c5f01e1dee0d41b8014ac6f1eeafda974917ba17

>The following VCPKG_OSX_SYSROOT values are currently supported:
> - iphoneos, iphonesimulator, or an absolute path to the device or
>   simulator Xcode SDK.

So the simulator is explicitly intended to work.

What does setting it break?

BillyONeal replied:
Also this whole block should already be disabled because `_vcpkg_ios_sysroot` isn't set above when arch is arm64.

colincornaby commented:
I've combined the toolchain files and it seems to be working in my testing so far. I left a few questions on the PR.

colincornaby commented:
I'm promoting this draft to a ready for review PR. I think my architecture questions still stand. I think they can be addressed during review.

BillyONeal commented on scripts/cmake/vcpkg_configure_cmake.cmake:
```diff
@@ -75,6 +75,8 @@ function(z_vcpkg_select_default_vcpkg_chainload_toolchain)
     elseif(VCPKG_TARGET_IS_IOS)
         set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/ios.cmake")
+    elseif(VCPKG_TARGET_IS_VISIONOS)
+        set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/visionos.cmake")
```
That this is broken scares me that this whole thing is untested.

colincornaby replied:
I checked in a correction and I'm looking at what happened here. I do have a test project using this that has valid output. And if I check the loader information on vcpkg it's correct, indicating it was build with the correct toolchain:

```
Load command 1
      cmd LC_BUILD_VERSION
  cmdsize 24
 platform 11
    minos 2.5
      sdk 2.5
   ntools 0
```

I'm blocked from running the full project on visionOS because some packages such as OpenSSL have their own platform checks which don't include visionOS - so I haven't been able to verify all dependencies for our application. So I've only been able to verify some packages are building and linking for visionOS.

The reference to the old toolchain does produce an error for me - and correcting it caused everything to compile.

BillyONeal commented on scripts/toolchains/ios.cmake:
```diff
@@ -29,7 +33,10 @@ if(NOT _VCPKG_IOS_TOOLCHAIN)
         unset(_vcpkg_ios_target_architecture)
 
-        if ("${arch}" STREQUAL "arm64")
+        if ("${_VCPKG_VISIONOS_TOOLCHAIN}" STREQUAL "visionOS")
+            set(_vcpkg_ios_system_processor "aarch64")
+            set(_vcpkg_ios_target_architecture "arm64")
+        elseif ("${arch}" STREQUAL "arm64")
```
As an aside, do you know why `${arch}` is wrong here that forced a branch to be added here?

colincornaby replied:
I actually don't think the if comparison is necessary - or actually correct.  😕 I'm removing the check entirely and letting it fall into the arm64 case. Seems to work just fine here.

BillyONeal reviewed (commented):
As far as I can see you should be able to revert all changes to `ios.cmake`?

colincornaby committed:
Combining iOS and visionOS toolchain files

colincornaby commented:
I think I've addressed the feedback in the cmake script. I want to verify the vcpkg outputs before moving on. I thought I ran the output on real hardware, but it's been long enough now that I'm unsure. Our full project is gated by several packages that have their own blocks on visionOS like OpenSSL. So I have to set up dedicated testing projects for visionOS.

Once I've done another pass and verified output on hardware successfully I'll re-promote the PR.

BillyONeal reviewed (approved):
This LGTM pending some confirmation that you tried a real example. (And expect to get tagged if it ever doesn't work because we aren't visionOS folks @colincornaby  😅)

BillyONeal reviewed (commented):
@vicroms @ras0219-msft @JavierMatosD and I discussed this today and we think adding the triplet is OK pending hardware test.

colincornaby commented:
I tested tonight and libpng output from vcpkg successfully linked - and I was able to call it's version function which returned the proper version.

If there are any issues with the visionOS support in the future I'd be happy to be tagged/take a look at issues. I'm working on getting a game engine ported over to visionOS that uses vcpkg as its dependency manager - so I'll be actively using this triplet.

Some packages like OpenSSL have platform in their builds that fail with OpenSSL - so this new triplet won't build everything. In cases like that I'm seeing PRs have been opened with some of those projects but are just pending acceptance.

BillyONeal commented:
> Some packages like OpenSSL have platform in their builds that fail with OpenSSL - so this new triplet won't build everything. In cases like that I'm seeing PRs have been opened with some of those projects but are just pending acceptance.

Sure, we don't need the whole registry to work, *a* thing that works is OK

BillyONeal commented:
> I tested tonight and libpng output from vcpkg successfully linked - and I was able to call it's version function which returned the proper version.

To clarify, this worked on the actual hardware?
